docs/guides/configuring-custom-entities.md

```diff
@@ -0,0 +1,169 @@
+# Configuring Custom Entities
+
+This is an **advanced** level guide for users using custom entities in Kong.
+Most users do need to use this feature.
+
+Kong has in-built extensibility with its plugin architecture.
+Plugins in Kong have a `config` property where users can store configuration
+for any custom plugin and this suffices in most use cases.
+In some use-cases, custom entities are created in Kong via plugins
+which are used to store additional configuration.
+This guide elaborates on how such custom entities can be used Kong Ingress
+Controller.
+
+> Note: All entities shipped with Kong are supported by Kong Ingress Controller
+out of the box. This guide applies only if you have a custom entity in your
+plugin. Custom plugins have first-class support in Kong Ingress Controller
+via the `KongPlugin` CRD. Please read [this](../setting-up-custom-plugins.md)
+guide for details.
+
+## Caveats
+
+- The feature discussed in this guide apply for DB-less deployments of Kong.
+  The feature is not supported for deployments where Kong is used with a
+  database or Kong is used in hybrid mode.
+  For these deployments, it is recommended to configure custom entities
+  directly using Kong's Admin API.
+- Custom entities which have a foreign relation with other core entities in Kong
+  are not supported. Only entities which can exist by themselves and then
+  be referenced via plugin configuration are supported.
+
+## Creating a JSON representation of the custom entity
+
+In this section, we will learn how to create a JSON representation of
+a custom entity.
+
+Suppose you have a custom entity in Kong with the following schema in Kong:
+
+```lua 
+{
+  name                = "xkcds",
+  primary_key         = { "id" },
+  cache_key           = { "name" },
+  endpoint_key        = "name",
+  fields = {
+    { id = typedefs.uuid },
+    {
+      name = {
+        type= "string",
+        required = true,
+        unique = true,
+      },
+    },
+    {
+      url = {
+        type = "string",
+        required = true,
+      },
+    },
+    { created_at = typedefs.auto_timestamp_s },
+    { updated_at = typedefs.auto_timestamp_s },
+  },
+}
+```
+
+An instance of such an entity would like:
+
+```json
+{
+  "id": "385def6e-3059-4929-bb12-d205e97284c5",
+  "name": "Bobby Drop Tables",
+  "url": "https://xkcd.com/327/"
+}
+```
+
+Multiple instances of such an entity are represented as follows:
+
+```json
+{
+  "xkcds": [
+    {
+      "id": "385def6e-3059-4929-bb12-d205e97284c5",
+      "name": "bobby_tables",
+      "url": "https://xkcd.com/327/"
+    },
+    {
+      "id": "d079a632-ac8d-4a9a-860c-71de82e8fc11",
+      "name": "compiling",
+      "url": "https://xkcd.com/303/"
+    }
+  ]
+}
+```
+
+If you have more than one custom entities that you would like to configure
+then you can create other entities by specifying the entity name at the root
+level of the JSON and then list the entities as a JSON array.
```

Comment:
This seems to imply that you can provide configuration without using entity names mapped to arrays of that entities--can you do so? I don't see a way it's possible, since there's no other way to indicate what type of entity exists in the secret.

In either case, I'd recommend including a second entity type in the example to show the format in practice. The schema for the additional entity should be omitted for brevity.

Reply:
I reworded a little bit to avoid adding another entity (this doc is already pretty hard to grok).